### PR TITLE
Handle empty string JSON response to DELETE request

### LINKED ISSUE #55: Empty string returned from API causes JSON error
Version 1.2.4
Sending a DELETE request to the `Item/<uid>` endpoint returns an empty `200` response. This results in an error when the `Manager` tries to return JSON. (myob/managers.py line 92):

```
response.headers.get('content-type')
#=> application/json
response.json()
#=> *** simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
response.content
#=> b''
```

The problem is the API, since it specifies the content as `application/json` yet returns invalid JSON. But maybe the package can handle this somehow?

### PR DESCRIPTION
**Pull request by jpmoral-factory**
Fixes #55

### CODE REVIEW

File: tests/test_endpoints.py [hunk 1]
```diff
@@ -54,6 +55,26 @@ def assertEndpointReached(self, func, params, method, endpoint, mock_request, ti
...
+        func(**params)
+        full_endpoint = 'https://api.myob.com/accountright' + endpoint
+        mock_request.assert_called_once_with(
+            'DELETE',
+            full_endpoint,
+            headers=self.expected_request_headers,
+            params={},
+            **({}),
+            timeout=timeout,
+        )
```
Comment (jarekwg) [inline tests/test_endpoints.py:76]: This isn't quite right.
Our intention is to ensure that if calling `response.json()` raises a JSONDecodeError, it gets handled gracefully and we return empty dict instead.

It seems you're just testing what the request function gets called with.

File: tests/test_endpoints.py [hunk 2]
```diff
@@ -378,6 +407,7 @@ def test_purchase_bills(self):
         self.assertEndpointReached(self.companyfile.purchase_bills.put_miscellaneous, {'uid': UID, 'data': DATA}, 'PUT', f'/{CID}/Purchase/Bill/Miscellaneous/{UID}/')
         self.assertEndpointReached(self.companyfile.purchase_bills.post_miscellaneous, {'data': DATA}, 'POST', f'/{CID}/Purchase/Bill/Miscellaneous/')
         self.assertEndpointReached(self.companyfile.purchase_bills.delete_miscellaneous, {'uid': UID}, 'DELETE', f'/{CID}/Purchase/Bill/Miscellaneous/{UID}/')
+        self.assertEmptyDeleteResponseHandled(self.companyfile.purchase_bills.delete_miscellaneous, {'uid': UID}, f'/{CID}/Purchase/Bill/Miscellaneous/{UID}/')
```
Comment (jarekwg) [inline tests/test_endpoints.py:410]: There's no need to add this test for multiple endpoints -- we're mocking the endpoints anyway.

**Review by jarekwg — CHANGES_REQUESTED** [review]

File: tests/test_endpoints.py (hunk 1 above)
Comment (jpmoral-factory) [reply tests/test_endpoints.py:76]: Well, not quite. With the old code the test blows up. But yes, I'll add a check for the empty dict return.

File: tests/test_endpoints.py (hunk 2 above)
Comment (jpmoral-factory) [reply tests/test_endpoints.py:410]: My first approach was to tack something onto `assertEndpointReached` to check that it handles empty strings from  `DELETE` requests. What do you suggest?

Comment (jarekwg) [reply tests/test_endpoints.py]: I think the `assertEndpointReached ` is chiefly interested in testing that we generate out the endpoint urls correctly.

So I'd be testing the return value separately.

File: tests/test_endpoints.py [hunk 3]
```diff
@@ -71,6 +72,18 @@ def test_base(self):
         self.expected_request_headers['x-myobapi-cftoken'] = ''
         self.assertEndpointReached(self.myob.info, {}, 'GET', '/Info/')
 
+    @patch('myob.managers.requests.request')
+    def test_empty_delete_response(self, mock_request):
+        mock_request.return_value.status_code = 200
+
+        def response_json():
+            return Response().json()
```
Comment (jarekwg) [inline tests/test_endpoints.py:80]: Better to explicitly be raising a `JSONDecodeError` error here -- that's what we want to be testing for.

**Review by jarekwg — CHANGES_REQUESTED** [review]
For completeness, we'd probs want to test the converse cases as well, that non-delete methods, or return values that aren't `b''`` still error.
But I'm not too fussed there.
Just left one more comment.

Thanks for contributing! :)

**Comment by jpmoral-factory** [comment]
> Thanks for contributing! :)
Thanks for creating/maintaining the project!

**Review by jarekwg — APPROVED** [review]

**Comment by jpmoral-factory** [comment]
:heart:

**Comment by jarekwg** [comment]
Out in [1.2.13](https://github.com/uptick/pymyob/releases/tag/1.2.13).

**Pull request merged**